crates/sui-indexer-alt/src/pipeline/concurrent/mod.rs

```diff
@@ -3,59 +3,78 @@
 
 use std::sync::Arc;
 
-use committer::committer;
 use sui_types::full_checkpoint_content::CheckpointData;
 use tokio::{sync::mpsc, task::JoinHandle};
 use tokio_util::sync::CancellationToken;
-use watermark::watermark;
 
 use crate::{
     db::Db, handlers::Handler, metrics::IndexerMetrics, models::watermarks::CommitterWatermark,
 };
 
-use super::{processor::processor, PipelineConfig, COMMITTER_BUFFER};
+use super::{processor::processor, PipelineConfig, PIPELINE_BUFFER};
 
+use self::{collector::collector, committer::committer, watermark::watermark};
+
+mod collector;
 mod committer;
 mod watermark;
 
 /// Start a new concurrent (out-of-order) indexing pipeline served by the handler, `H`. Starting
 /// strictly after the `watermark` (or from the beginning if no watermark was provided).
 ///
 /// Each pipeline consists of a processor task which takes checkpoint data and breaks it down into
-/// rows, ready for insertion, and a committer which writes those rows out to the database.
+/// rows, ready for insertion, a collector which batches those rows into an appropriate size for
+/// the database, a committer which writes the rows out concurrently, and a watermark task to
+/// update the high watermark.
+///
 /// Committing is performed out-of-order: the pipeline may write out checkpoints out-of-order,
 /// either because it received the checkpoints out-of-order or because of variance in processing
 /// time.
 ///
-/// The committer also maintains a row in the `watermarks` table for the pipeline which tracks the
+/// The pipeline also maintains a row in the `watermarks` table for the pipeline which tracks the
 /// watermark below which all data has been committed (modulo pruning).
 ///
-/// Checkpoint data is fed into the pipeline through the `checkpoint_rx` channel, and an internal
-/// channel is created to communicate checkpoint-wise data to the committer. The pipeline can be
-/// shutdown using its `cancel` token.
+/// Checkpoint data is fed into the pipeline through the `checkpoint_rx` channel, and internal
+/// channels are created to communicate between its various components. The pipeline can be
+/// shutdown using its `cancel` token, and will also shutdown if any of its independent tasks
+/// reports an issue.
 pub fn pipeline<H: Handler + 'static>(
```

Comment:
I split it up because in the next PR I'm adding a `sequential` sub directory that performs the in-order commit (but it shares the same processor task).

Reply:
Why do we want to add that right now?

Reply:
The objects pipelines have to be sequential, so it's for them.